traffic_ops/traffic_ops_golang/server/servers.go

```diff
@@ -225,15 +264,77 @@ func getServers(params map[string]string, tx *sqlx.Tx, privLevel int) ([]tc.Serv
 		if err = rows.StructScan(&s); err != nil {
 			return nil, []error{fmt.Errorf("getting servers: %v", err)}, tc.SystemError
 		}
-		if privLevel < auth.PrivLevelAdmin {
+		if user.PrivLevel < auth.PrivLevelAdmin {
 			s.ILOPassword = &HiddenField
 			s.XMPPPasswd = &HiddenField
 		}
 		servers = append(servers, s)
 	}
+
+	// if ds requested uses mid-tier caches, add those to the list as well
+	if usesMids {
+		mids, errs, errType := getMidServers(servers, tx)
+
+		log.Debugf("getting mids: %v, %v\n", errs, errType)
+		if len(errs) > 0 {
+			for _, err := range errs {
+				if err.Error() == `id cannot parse to integer` {
+					return nil, []error{errors.New("Resource not found.")}, tc.DataMissingError //matches perl response
+				}
+			}
+			return nil, errs, errType
+		}
+		for _, server := range mids {
+			servers = append(servers, server)
+		}
+	}
+
 	return servers, nil, tc.NoError
 }
 
+// getMidServers gets mids used by the servers in this ds
+// Original comment from the Perl code:
+// if the delivery service employs mids, we're gonna pull mid servers too by pulling the cachegroups of the edges and finding those cachegroups parent cachegroup...
+// then we see which servers have cachegroup in parent cachegroup list...that's how we find mids for the ds :)
+func getMidServers(servers []tc.ServerNullable, tx *sqlx.Tx) ([]tc.ServerNullable, []error, tc.ApiErrorType) {
+	if len(servers) == 0 {
+		return nil, nil, tc.NoError
+	}
+	var ids []string
+	for _, s := range servers {
+		ids = append(ids, strconv.Itoa(*s.ID))
+	}
+
+	edgeIDs := strings.Join(ids, ",")
+	// TODO: include secondary parent?
```

Comment:
this TODO something to worry about?

Reply:
no -- Perl version does not refer to secondary parent at all